File: app/src/main/java/de/christian2003/smarthome/view/room/RoomViewModel.java
package de.christian2003.smarthome.view.room;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModel;

import java.util.ArrayList;

import de.christian2003.smarthome.model.data.ShRoom;
import de.christian2003.smarthome.model.user_information.InformationTitle;
import de.christian2003.smarthome.model.user_information.InformationType;
import de.christian2003.smarthome.model.user_information.UserInformation;


public class RoomViewModel extends ViewModel {

    @Nullable
    private ShRoom room;

    private final ArrayList<UserInformation> warnings;

    private final ArrayList<UserInformation> errors;


    public RoomViewModel() {
        room = null;
        warnings = new ArrayList<>();
        errors = new ArrayList<>();
    }


    @Nullable
    public ShRoom getRoom() {
        return room;
    }

    public void setRoom(@NonNull ShRoom room) {
        this.room = room;
        warnings.clear();
        errors.clear();
        for (UserInformation userInformation : room.getUserInformation()) {
            if (userInformation.getInformationType() == InformationType.WARNING) {
                warnings.add(userInformation);
            }
            else if (userInformation.getInformationType() == InformationType.ERROR) {
                errors.add(userInformation);
            }
        }
    }

    @NonNull
    public ArrayList<UserInformation> getWarnings() {
        return warnings;
    }

    @NonNull
    public ArrayList<UserInformation> getErrors() {
        return errors;
    }

}

File: common/data/src/main/java/de/christian2003/smarthome/data/view/room/RoomViewModel.kt
package de.christian2003.smarthome.data.view.room

import android.app.Application
import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.AndroidViewModel
import de.christian2003.smarthome.data.model.SmartHomeRepository
import de.christian2003.smarthome.data.model.room.ShRoom


/**
 * Class implements the view model for the page to display the contents of a room.
 */
class RoomViewModel(application: Application): AndroidViewModel(application) {

    /**
     * Attribute stores the preferences in which to store the URL.
     */
    private val preferences = getApplication<Application>().getSharedPreferences("smart_home", Context.MODE_PRIVATE)

    /**
     * Attribute stores the repository through which to access the data.
     */
    private lateinit var repository: SmartHomeRepository

    /**
     * Attribute stores whether to display warnings.
     */
    var showWarnings: Boolean by mutableStateOf(true)

    /**
     * Attribute stores whether to display errors.
     */
    var showErrors: Boolean by mutableStateOf(true)

    /**
     * Attribute stores the room to display.
     */
    lateinit var room: ShRoom

    /**
     * Attribute stores the items of the room that shall be displayed. This contains user information,
     * infos as well as devices.
     */
    lateinit var items: List<Any>


    /**
     * Method initializes the view model.
     *
     * @param repository    Repository from which to source the data.
     * @param position      Position of the room to display.
     */
    fun init(repository: SmartHomeRepository, position: Int) {
        this.repository = repository
        room = repository.rooms[position]

        showWarnings = preferences.getBoolean("show_warnings", true)
        showErrors = preferences.getBoolean("show_errors", true)

        val mutableList = mutableListOf<Any>()
        mutableList.addAll(room.userInformation)
        mutableList.addAll(room.infos)
        mutableList.addAll(room.devices)
        items = mutableList
    }

}
